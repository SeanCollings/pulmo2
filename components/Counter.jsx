import React, { useState, useEffect, useRef } from 'react';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  Animated,
  View,
} from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import { useTheme } from '../hooks/useTheme';

const firstNumberIs1 = (number) => {
  if (number > 99) return false;

  const regEx = /^1[0-9].*$/;
  return regEx.test(number.toString()) || number === 1;
};

const Counter = ({
  reset,
  disabled,
  duration = 200,
  countStart = 0,
  jest = false,
}) => {
  const theme = useTheme();
  const firstMount = useRef(true);
  const [count, setCount] = useState(countStart);
  const [animatedRotate, setAnimatedRotate] = useState(new Animated.Value(0));
  const [animatedGrow] = useState(new Animated.Value(0));
  const animatedOpacity = new Animated.Value(0);

  const interpolateRotate = animatedRotate.interpolate({
    inputRange: [0, 1],
    outputRange: ['-90deg', '0deg'],
  });

  const rotateAnimation = (duration) =>
    Animated.timing(animatedRotate, {
      toValue: 1,
      duration,
      delay: firstMount.current ? 100 : 0,
      useNativeDriver: false,
    });
  const opacityAnimation = Animated.timing(animatedOpacity, {
    toValue: 1,
    duration,
    useNativeDriver: false,
  });
  const growAnimation = Animated.timing(animatedGrow, {
    toValue: 1,
    duration,
    useNativeDriver: false,
  });

  useEffect(() => {
    Animated.parallel([
      growAnimation,
      rotateAnimation(duration),
      opacityAnimation,
    ]).start(({ finished }) => {
      if (finished && !jest) setAnimatedRotate(new Animated.Value(0));
    });
  }, []);

  useEffect(() => {
    if (firstMount.current) {
      firstMount.current = false;
    } else if (reset) {
      rotateAnimation(duration).start(({ finished }) => {
        if (finished && !jest) setAnimatedRotate(new Animated.Value(0));
      });
      setCount(0);
    }
  }, [reset]);

  const countUpHandler = () => {
    if (count < 999) setCount((curr) => curr + 1);
  };
  const countDownHandler = () => {
    if (count > 0) setCount((curr) => curr - 1);
  };

  const animatedRotateStyle = {
    transform: [{ rotate: interpolateRotate }],
  };
  const animatedGrowStyle = {
    transform: [{ scale: animatedGrow }],
  };

  return (
    <Animated.View
      style={{
        ...styles.container,
        opacity: animatedOpacity,
        backgroundColor:
          firstMount.current || disabled ? theme.BACKGROUND : theme.SECONDARY,
      }}
    >
      <View style={{ opacity: disabled ? 0.5 : 1 }}>
        <Animated.View style={{ ...animatedGrowStyle }}>
          <TouchableOpacity
            testID="counterButton"
            activeOpacity={0.8}
            style={{
              ...styles.fabContainer,
              backgroundColor: theme.BACKGROUND,
              borderColor: theme.SECONDARY,
            }}
            onPress={countUpHandler}
            onLongPress={countDownHandler}
            disabled={disabled}
          >
            <Animated.View style={{ ...animatedRotateStyle }}>
              {count === 0 && (
                <MaterialCommunityIcons
                  name={'plus'}
                  color={theme.SECONDARY}
                  size={30}
                />
              )}
              {count > 0 && (
                <Text
                  style={{
                    ...styles.textStyle,
                    color: theme.SECONDARY,
                    marginRight: firstNumberIs1(count) ? 3 : 0,
                  }}
                >
                  {count}
                </Text>
              )}
            </Animated.View>
          </TouchableOpacity>
        </Animated.View>
      </View>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    position: 'absolute',
    bottom: 0,
    right: 0,
    marginRight: 15,
    marginBottom: 20,
    borderRadius: 55,
  },
  fabContainer: {
    width: 55,
    height: 55,
    borderRadius: 55,
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 2,
  },
  textStyle: {
    textAlign: 'center',
    fontFamily: 'tit-regular',
    fontSize: 30,
    marginTop: -5,
  },
});

export default Counter;
