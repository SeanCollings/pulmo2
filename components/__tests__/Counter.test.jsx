import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import Counter from '../Counter';

describe('Counter - unit test', () => {
  let component;

  afterEach(() => {
    cleanup();
  });

  test('should render in disabled state', async () => {
    await act(async () => {
      component = create(<Counter disabled jest duration={0} />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should render in active state', async () => {
    await act(async () => {
      component = create(<Counter jest duration={0} />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should increase counter', async () => {
    await act(async () => {
      component = create(<Counter jest duration={0} />);
    });

    const counterButton = component.root.find(
      ({ props }) => props.testID === 'counterButton'
    );

    fireEvent.press(counterButton);
    expect(component).toMatchSnapshot();
  });

  test('should increase counter', async () => {
    await act(async () => {
      component = create(<Counter jest duration={0} />);
    });

    const counterButton = component.root.find(
      ({ props }) => props.testID === 'counterButton'
    );

    fireEvent.press(counterButton);
    expect(component).toMatchSnapshot();
  });

  test('should decrease counter on longPress', async () => {
    await act(async () => {
      component = create(<Counter jest countStart={10} duration={0} />);
    });

    const counterButton = component.root.find(
      ({ props }) => props.testID === 'counterButton'
    );

    fireEvent(counterButton, 'onLongPress');
    expect(component).toMatchSnapshot();
  });
});
