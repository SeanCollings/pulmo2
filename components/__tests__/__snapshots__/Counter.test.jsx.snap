// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter - unit test should decrease counter on longPress 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#002f56",
      "borderRadius": 55,
      "bottom": 0,
      "marginBottom": 20,
      "marginRight": 15,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#002f56",
            "borderRadius": 55,
            "borderWidth": 2,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 55,
          }
        }
        testID="counterButton"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "-90deg",
                },
              ],
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#002f56",
                "fontFamily": "tit-regular",
                "fontSize": 30,
                "marginRight": 0,
                "marginTop": -5,
                "textAlign": "center",
              }
            }
          >
            9
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Counter - unit test should increase counter 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#002f56",
      "borderRadius": 55,
      "bottom": 0,
      "marginBottom": 20,
      "marginRight": 15,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#002f56",
            "borderRadius": 55,
            "borderWidth": 2,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 55,
          }
        }
        testID="counterButton"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "-90deg",
                },
              ],
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#002f56",
                "fontFamily": "tit-regular",
                "fontSize": 30,
                "marginRight": 3,
                "marginTop": -5,
                "textAlign": "center",
              }
            }
          >
            1
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Counter - unit test should increase counter 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#002f56",
      "borderRadius": 55,
      "bottom": 0,
      "marginBottom": 20,
      "marginRight": 15,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#002f56",
            "borderRadius": 55,
            "borderWidth": 2,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 55,
          }
        }
        testID="counterButton"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "-90deg",
                },
              ],
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#002f56",
                "fontFamily": "tit-regular",
                "fontSize": 30,
                "marginRight": 3,
                "marginTop": -5,
                "textAlign": "center",
              }
            }
          >
            1
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Counter - unit test should render in active state 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 55,
      "bottom": 0,
      "marginBottom": 20,
      "marginRight": 15,
      "opacity": 1,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#002f56",
            "borderRadius": 55,
            "borderWidth": 2,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 55,
          }
        }
        testID="counterButton"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "-90deg",
                },
              ],
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#002f56",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "material-community",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Ôêï
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Counter - unit test should render in disabled state 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 55,
      "bottom": 0,
      "marginBottom": 20,
      "marginRight": 15,
      "opacity": 1,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    style={
      Object {
        "opacity": 0.5,
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#002f56",
            "borderRadius": 55,
            "borderWidth": 2,
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 55,
          }
        }
        testID="counterButton"
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "-90deg",
                },
              ],
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#002f56",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "material-community",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Ôêï
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
